helper/metricsutil/gauge_process.go

```diff
@@ -0,0 +1,236 @@
+		id,
+		collector,
+		logger,
+		defaultClock{},
+	)
+}
+
+// test version allows an alternative clock implementation
+func (m *ClusterMetricSink) newGcpWithClock(
+	key []string,
+	id []Label,
+	collector GaugeCollector,
+	logger log.Logger,
+	clock clock,
+) (*GaugeCollectionProcess, error) {
+	process := &GaugeCollectionProcess{
+		Stop:             make(chan struct{}, 1),
+		stopped:          make(chan struct{}, 1),
+		key:              key,
+		labels:           id,
+		collector:        collector,
+		sink:             m,
+		originalInterval: m.GaugeInterval,
+		currentInterval:  m.GaugeInterval,
+		logger:           logger,
+		clock:            clock,
+	}
+	return process, nil
+}
+
+// delayStart randomly delays by up to one extra interval
+// so that collection processes do not all run at the time time.
+// If we knew all the procsses in advance, we could just schedule them
+// evenly, but a new one could be added per secret engine.
+func (p *GaugeCollectionProcess) delayStart() bool {
+	randomDelay := time.Duration(rand.Intn(int(p.currentInterval)))
+	// FIXME: a Timer might be better, but then we'd have to simulate
+	// one of those too?
+	delayTick := p.clock.NewTicker(randomDelay)
+	defer delayTick.Stop()
+
+	select {
+	case <-p.Stop:
+		return true
+	case <-delayTick.C:
+		break
+	}
+	return false
+}
+
+// resetTicker stops the old ticker and starts a new one at the current
+// interval setting.
+func (p *GaugeCollectionProcess) resetTicker() {
+	if p.ticker != nil {
+		p.ticker.Stop()
+	}
+	p.ticker = p.clock.NewTicker(p.currentInterval)
+}
+
+// collectAndFilterGauges executes the callback function,
+// limits the cardinality, and streams the results to the metrics sink.
+func (p *GaugeCollectionProcess) collectAndFilterGauges() {
+	// Run for only an allotted amount of time.
+	timeout := time.Duration(collectionBound * float64(p.currentInterval))
+	ctx, cancel := context.WithTimeout(context.Background(),
+		timeout)
+	defer cancel()
+
+	start := p.clock.Now()
+	values, err := p.collector(ctx)
+	end := p.clock.Now()
+	duration := end.Sub(start)
+
+	// Report how long it took to perform the operation.
+	p.sink.AddDurationWithLabels([]string{"metrics", "collection"},
+		duration,
+		p.labels)
+
+	// If over threshold, back off by doubling the measurement interval.
+	// Currently a restart is the only way to bring it back down.
+	threshold := time.Duration(collectionTarget * float64(p.currentInterval))
+	if duration > threshold {
+		p.logger.Warn("gauge collection time exceeded target", "target", threshold, "actual", duration, "id", p.labels)
+		p.currentInterval *= 2
+		p.resetTicker()
+	}
+
+	if err != nil {
+		p.logger.Error("error collecting gauge", "id", p.labels, "error", err)
+		return
+	}
+
+	// Filter to top N.
+	// This does not guarantee total cardinality is <= N, but it does slow things down
+	// a little if the cardinality *is* too high and the gauge needs to be disabled.
+	if len(values) > p.sink.MaxGaugeCardinality {
```

Comment:
Would a metric reporting the actual number of values before filtering be useful?

Reply:
I think it would be better to report on the number of whatever is the actual underlying source of data, like the number of mount points-- which will come later.